kex-core/src/main/kotlin/org/vorpal/research/kex/asserter/ExecutionFinalInfo.kt

```diff
@@ -0,0 +1,97 @@
+package org.vorpal.research.kex.asserter
+
+import org.vorpal.research.kex.descriptor.ConstantDescriptor
+import org.vorpal.research.kex.descriptor.Descriptor
+import org.vorpal.research.kex.descriptor.DescriptorBuilder
+import org.vorpal.research.kex.descriptor.ObjectDescriptor
+import org.vorpal.research.kex.ktype.KexClass
+import org.vorpal.research.kex.parameters.Parameters
+import org.vorpal.research.kex.state.term.ArgumentTerm
+import org.vorpal.research.kex.trace.symbolic.protocol.ExceptionResult
+import org.vorpal.research.kex.trace.symbolic.protocol.ExecutionCompletedResult
+import org.vorpal.research.kex.trace.symbolic.protocol.ExecutionResult
+import org.vorpal.research.kex.trace.symbolic.protocol.SuccessResult
+import org.vorpal.research.kfg.ir.Method
+import org.vorpal.research.kfg.ir.value.Constant
+import org.vorpal.research.kfg.ir.value.instruction.ReturnInst
+import org.vorpal.research.kfg.type.BoolType
+import org.vorpal.research.kthelper.logging.log
+import org.vorpal.research.kthelper.toBoolean
+
+sealed class ExecutionFinalInfo<T>() {
+    abstract val instance: T?
+    abstract val args: List<T>
+    fun isException(): Boolean = this is ExecutionExceptionFinalInfo
+}
+
+class ExecutionSuccessFinalInfo<T>(
+    override val instance: T?,
+    override val args: List<T>,
+    val retValue: T?
+) : ExecutionFinalInfo<T>()
+
+class ExecutionExceptionFinalInfo<T>(
+    override val instance: T?,
+    override val args: List<T>,
+    val javaClass: String
+): ExecutionFinalInfo<T>()
+
+fun Parameters<Descriptor>.extractExceptionFinalInfo(exceptionJavaName: String): ExecutionExceptionFinalInfo<Descriptor> =
+    ExecutionExceptionFinalInfo(instance, arguments, exceptionJavaName)
+
+fun Parameters<Descriptor>.extractSuccessFinalInfo(returnValueDescriptor: Descriptor?) =
+    ExecutionSuccessFinalInfo(instance, arguments, returnValueDescriptor)
+
+fun extractFinalInfo(executionResult: ExecutionResult, method: Method): ExecutionFinalInfo<Descriptor>? = when (executionResult) {
+    is ExecutionCompletedResult -> {
+        val instance = executionResult.symbolicState.concreteValues.entries
+            .firstOrNull { it.key.name == "this" }?.value?.deepCopy()
+        val args = executionResult.symbolicState.concreteValues
+            .filterKeys { it is ArgumentTerm }.map { it.value.deepCopy() }
```

Comment:
Why are we deep-copying everything?

Reply:
If I remember correctly if we will not use deep copy the initialisation of variables for oracles will have the same fields as the arguments. In other words, if class A has a field "list" with list reference and we will not make a deep copy, when both variables arg and arg_copy will have the same value in the "list" field. I may be mistaken, I will recheck again